### PR TITLE
fh convert: init

### PR DESCRIPTION
**Pull request by cole-h**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -36,9 +37,7 @@ serde = { version = "1.0.188", default-features = false, features = ["derive"] }
 serde_json = "1.0.105"
 thiserror = { version = "1.0.44", default-features = false }
 tokio = { version = "1.30.0", default-features = false, features = [
-  "fs",
-  "macros",
-  "rt-multi-thread",
+  "full",
```
Comment (cole-h) [inline Cargo.toml:40]: We pull in the same amount of deps whether this is trimmed down to the previous list, or the "full" set.

**Review by grahamc — APPROVED** [review]
I've lightly tested this against a few examples, but not against all the acceptance criteria of the ticket. I've also not really reviewed the code, but I'm so hyped for this that I'm throwing an approve on it.

**Review by grahamc — APPROVED** [review]
This is doing exactly what I hoped.

I tested with this:

```nix
{
  inputs = {
    nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    nixos-unstable-small.url = "github:NixOS/nixpkgs/nixos-unstable-small";
    nixos-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";

    nixpkgs-23-05-darwin.url = "github:NixOS/nixpkgs/nixpkgs-23.05-darwin";
    nixos-23-05-small.url = "github:NixOS/nixpkgs/nixos-23.05-small";
    nixos-23-05.url = "github:NixOS/nixpkgs/nixos-23.05";

    nixpkgs-22-11-darwin.url = "github:NixOS/nixpkgs/nixpkgs-22.11-darwin";
    nixos-22-11-small.url = "github:NixOS/nixpkgs/nixos-22.11-small";
    nixos-22-11.url = "github:NixOS/nixpkgs/nixos-22.11";

    lower-nixpkgs-unstable.url = "github:nixos/nixpkgs/nixpkgs-unstable";
    lower-nixos-unstable-small.url = "github:nixos/nixpkgs/nixos-unstable-small";
    lower-nixos-unstable.url = "github:nixos/nixpkgs/nixos-unstable";

    lower-nixpkgs-23-05-darwin.url = "github:nixos/nixpkgs/nixpkgs-23.05-darwin";
    lower-nixos-23-05-small.url = "github:nixos/nixpkgs/nixos-23.05-small";
    lower-nixos-23-05.url = "github:nixos/nixpkgs/nixos-23.05";

    lower-nixpkgs-22-11-darwin.url = "github:nixos/nixpkgs/nixpkgs-22.11-darwin";
    lower-nixos-22-11-small.url = "github:nixos/nixpkgs/nixos-22.11-small";
    lower-nixos-22-11.url = "github:nixos/nixpkgs/nixos-22.11";


    spongebob-nixpkgs-unstable.url = "github:NiXoS/nIxPkGs/nixpkgs-unstable";
    spongebob-nixos-unstable-small.url = "github:NiXoS/nIxPkGs/nixos-unstable-small";
    spongebob-nixos-unstable.url = "github:NiXoS/nIxPkGs/nixos-unstable";

    spongebob-nixpkgs-23-05-darwin.url = "github:NiXoS/nIxPkGs/nixpkgs-23.05-darwin";
    spongebob-nixos-23-05-small.url = "github:NiXoS/nIxPkGs/nixos-23.05-small";
    spongebob-nixos-23-05.url = "github:NiXoS/nIxPkGs/nixos-23.05";

    spongebob-nixpkgs-22-11-darwin.url = "github:NiXoS/nIxPkGs/nixpkgs-22.11-darwin";
    spongebob-nixos-22-11-small.url = "github:NiXoS/nIxPkGs/nixos-22.11-small";
    spongebob-nixos-22-11.url = "github:NiXoS/nIxPkGs/nixos-22.11";

    registry-nixpkgs.url = "nixpkgs";

    registry-nixpkgs-unstable.url = "nixpkgs/nixpkgs-unstable";
    registry-nixos-unstable-small.url = "nixpkgs/nixos-unstable-small";
    registry-nixos-unstable.url = "nixpkgs/nixos-unstable";

    registry-nixpkgs-23-05-darwin.url = "nixpkgs/nixpkgs-23.05-darwin";
    registry-nixos-23-05-small.url = "nixpkgs/nixos-23.05-small";
    registry-nixos-23-05.url = "nixpkgs/nixos-23.05";

    registry-nixpkgs-22-11-darwin.url = "nixpkgs/nixpkgs-22.11-darwin";
    registry-nixos-22-11-small.url = "nixpkgs/nixos-22.11-small";
    registry-nixos-22-11.url = "nixpkgs/nixos-22.11";

    copier.url = "github:copier-org/copier/8.3.0";
    copier-fake.url = "github:copier-org/copier/9.0.0";
    copier-caps.url = "github:coPiEr-org/coPiEr/8.3.0";

    copier-caps-master.url = "github:coPiEr-org/coPiEr/master";

    flake-compat = {
      url = "github:edolstra/flake-compat";
      flake = false;
    };
  };
  outputs = { nixpkgs, ... } @ inputs: {};
}
```

and got this:

```nix
{
  inputs = {
    nixpkgs.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    nixpkgs-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    nixos-unstable-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    nixos-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";

    nixpkgs-23-05-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    nixos-23-05-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    nixos-23-05.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";

    nixpkgs-22-11-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    nixos-22-11-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    nixos-22-11.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";

    lower-nixpkgs-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    lower-nixos-unstable-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    lower-nixos-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";

    lower-nixpkgs-23-05-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    lower-nixos-23-05-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    lower-nixos-23-05.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";

    lower-nixpkgs-22-11-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    lower-nixos-22-11-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    lower-nixos-22-11.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";


    spongebob-nixpkgs-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    spongebob-nixos-unstable-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    spongebob-nixos-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";

    spongebob-nixpkgs-23-05-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    spongebob-nixos-23-05-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    spongebob-nixos-23-05.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";

    spongebob-nixpkgs-22-11-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    spongebob-nixos-22-11-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    spongebob-nixos-22-11.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";

    registry-nixpkgs.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";

    registry-nixpkgs-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    registry-nixos-unstable-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";
    registry-nixos-unstable.url = "https://flakehub.com/f/NixOS/nixpkgs/0.1.532072.tar.gz";

    registry-nixpkgs-23-05-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    registry-nixos-23-05-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";
    registry-nixos-23-05.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2305.491440.tar.gz";

    registry-nixpkgs-22-11-darwin.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    registry-nixos-22-11-small.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";
    registry-nixos-22-11.url = "https://flakehub.com/f/NixOS/nixpkgs/0.2211.433406.tar.gz";

    copier.url = "https://flakehub.com/f/copier-org/copier/8.3.0.tar.gz";
    copier-fake.url = "github:copier-org/copier/9.0.0";
    copier-caps.url = "https://flakehub.com/f/copier-org/copier/8.3.0.tar.gz";

    copier-caps-master.url = "github:coPiEr-org/coPiEr/master";

    flake-compat.url = "https://flakehub.com/f/edolstra/flake-compat/1.0.1.tar.gz";
  };
  outputs = { nixpkgs, ... } @ inputs: {};
}

```

File: src/cli/cmd/convert.rs [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+                } else {
+                    tokio::fs::write(shell_nix_path, contents).await?;
+                }
+            }
+        }
+
+        if default_nix_path.exists() {
+            let existing_contents = tokio::fs::read_to_string(&default_nix_path).await?;
+            if existing_contents.contains(FLAKE_COMPAT_MARKER) {
+                let contents = format!("{FLAKE_COMPAT_CONTENTS_PREFIX}.defaultNix");
```
Comment (grahamc) [inline src/cli/cmd/convert.rs:366]:
```suggestion
                let contents = format!("{FLAKE_COMPAT_CONTENTS_PREFIX}.defaultNix\n");
```

File: src/cli/cmd/convert.rs [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+                if line.contains("default.nix") {
+                    default_nix_clean = false;
+                }
+            }
+        }
+
+        if shell_nix_path.exists() {
+            let existing_contents = tokio::fs::read_to_string(&shell_nix_path).await?;
+            if existing_contents.contains(FLAKE_COMPAT_MARKER) {
+                let contents = format!("{FLAKE_COMPAT_CONTENTS_PREFIX}.shellNix");
```
Comment (grahamc) [inline src/cli/cmd/convert.rs:351]:
```suggestion
                let contents = format!("{FLAKE_COMPAT_CONTENTS_PREFIX}.shellNix\n");
```

**Review by grahamc — APPROVED** [review]
I made a few edits in this PR, but then it got a bit too complicated for inline review ... so I made a PR: https://github.com/DeterminateSystems/fh/pull/54

**Pull request merged**